### PR TITLE
ARROW-12443: [C++][Gandiva] Implement castVARCHAR function for varbinary input

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement castVARCHAR function for varbinary input

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -534,88 +534,96 @@ const char* castVARCHAR_bool_int64(gdv_int64 context, gdv_boolean value,
...
+      *out_length = 0;                                                                 \
+      return "";                                                                       \
+    }                                                                                  \
+                                                                                       \
+    *out_length = index + byte_pos;                                                    \
+    return data;                                                                       \
+  }
+
+CAST_VARCHAR_UTF8_BINARY(utf8)
+CAST_VARCHAR_UTF8_BINARY(binary)
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:624]: castVARBINARY shouldn't need utf8 handling. Thats only for varchar.

Comment (jpedroantunes) [reply cpp/src/gandiva/precompiled/string_ops.cc:624]: Sure! But this IS a castVARCHAR function receiving a binary value as input. Can you confirm that you understand correctly?

Comment (projjal) [reply cpp/src/gandiva/precompiled/string_ops.cc:624]: oh it makes sense. I understood incorrectly.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 2]
```diff
@@ -534,88 +534,96 @@ const char* castVARCHAR_bool_int64(gdv_int64 context, gdv_boolean value,
...
+    if (byte_pos < 0) {                                                                \
+      *out_length = 0;                                                                 \
+      return "";                                                                       \
+    }                                                                                  \
+                                                                                       \
+    *out_length = index + byte_pos;                                                    \
+    return data;                                                                       \
+  }
+
+CAST_VARCHAR_UTF8_BINARY(utf8)
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:623]: These macro names looks confusing. Can you name the macro CAST_VARCHAR_FROM_VARWIDTH(TYPE)

**Review by projjal — APPROVED** [review]

**Review by projjal — APPROVED** [review]

**Pull request closed without merging**